sam: atsamc21: driver for CAN for SAMC21

benys opened the pull request:
Driver was based on can_sam.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,181 @@
+	struct can_mcan_msg_sram msg_ram;
+};
+
+void can_sam0_line_x_isr(const struct device *dev) {
+	can_mcan_line_0_isr(dev);
+	can_mcan_line_1_isr(dev);
+}
```
Does the SAM0 series only use one M_CAN interrupt line?

benys replied:
Yes. SAM21 have only one interrupt line. But I saw that can_mcan is expected 2 lines -> I fire 2 ISRs

henrikbrixandersen reviewed (changes requested):
We will need a way of testing this - at least the compilation part. Can this be enabled on a board?

benys commented:
> We will need a way of testing this - at least the compilation part. Can this be enabled on a board?

This CAN works only with SAMCAN21x. I have in my solution connected to board intruduced at #50384 PR. 
In my opionion I should add connection to board/soc after merged #50384? Becouse only C21 implements this SAM0_CAN

BTW In my repo it fully works. You can check my old commit: https://github.com/benys/zephyr/commit/55559a875e6ac83e73927d861b03ecd1ca3ee757

henrikbrixandersen commented:
Okay. I suggest postponing this to after #50384 is merged, then. That was the initial feedback given in that PR as well.

alexanderwachter commented:
Please provide a meaningful commit message

str4t0m commented:
Converted to a draft for now, as board files still need to be added after #50384 was merged.

benys commented:
@henrikbrixandersen i have a problem, that CAN can send messages (i receive messages by CAN usb stick), but when I send by CAN usb stick - i don't receive interrupt.
I set this filter:
```
    const struct can_filter change_led_filter = {
		.id_type = CAN_EXTENDED_IDENTIFIER,
		.rtr = CAN_DATAFRAME,
		.rtr_mask = 1,
		.id_mask = 0
	};

    ret = can_add_rx_filter_msgq(can_dev, &can_msgq, &change_led_filter);
	if (ret == -ENOSPC) {
		printk("Error, no filter available!\n");
		return;
	}
```
Do you have idea what I have to check?

benys commented:
@henrikbrixandersen 
> * Have you tried using the CAN shell for testing?
Not yet, but I run sam counter sample. And I extend it a litte bit. On loopback mode it fully works. With external CAN commands it not handle incoming CAN frames.
> * Are the pinctrl settings correct?
I tripled checked. I honestly I port my application which originaly used plain ASF4. So I checked pinctrls by comparison one-to-one. TXpin works becouse it send data. SO i believe that in in zephyr RXpin also is corecrly connected.
> * Is the CAN interrupt enabled correctly?
Yes, but I have idea that problem is that zephyr mcan driver uses two interrupts per device, but SAMC has only one line. I tried investigate, that maybe mcan driver set to use second interrupts for incoming data from external world.
> * Is the CAN interrupt triggered at all?
Yes for example:
```
LED ON
[00:00:05.004,000] <dbg> can_mcan: can_mcan_send: Sending 2 bytes. Id: 0x12345, ID type: extended   
[00:00:05.004,000] <dbg> can_mcan: can_mcan_send: Waiting for TX complete
[00:00:05.005,000] <dbg> can_mcan: can_mcan_line_1_isr: RX FIFO0 INT
[00:00:05.005,000] <dbg> can_mcan: can_mcan_get_message: Frame on filter 28, ID: 0x12345
Counter received: 2
[00:00:06.005,000] <dbg> can_mcan: can_mcan_send: Sending 1 bytes. Id: 0x10, ID type: standard   
[00:00:06.006,000] <dbg> can_mcan: can_mcan_line_1_isr: RX FIFO0 INT
[00:00:06.006,000] <dbg> can_mcan: can_mcan_get_message: Frame on filter 0, ID: 0x10
```
> * Have you tried using a debugger to examine what happens?
Yes :-) without debugger it is inposible to investigate :)

I tried check with second interupt for bosch IP. If you have some ideas please give me :)

alexanderwachter commented:
Looks like the loopback mode works.
That means it is likely that it is either pinctrl, or a problem wit the bus (baudrate, termination, issue with the transceiver, ...)

benys commented:
@henrikbrixandersen  and @alexanderwachter  i think that i found I receive some of messages.
I forget that "if you work with can you must found mcan datasheet" :)
```
#ifdef CONFIG_CAN_STM32FD
	can->ils = CAN_MCAN_ILS_RXFIFO0 | CAN_MCAN_ILS_RXFIFO1;
#else
	can->ils = 0;//CAN_MCAN_ILS_RF0N | CAN_MCAN_ILS_RF1N;
#endif
	can->ile = CAN_MCAN_ILE_EINT0; //| CAN_MCAN_ILE_EINT1;
```
in mcan driver was implemented that some interrupts was handled by second line. 
So I have a question - you recomend add elif CONFIG_CAN_SAM0 ? with fixed declaration for SAMC?

BTW i need now investiage why have many bus_off state.

henrikbrixandersen commented:
> So I have a question - you recomend add elif CONFIG_CAN_SAM0 ?

You can add that for now. I am working on getting rid of all platform specific code in the M_CAN backend driver, though.

benys commented:
@henrikbrixandersen @alexanderwachter i had last problem (i hope). NBTP computation.
I discovered that zephry compute this registry as:

**generation from zephyr**
- ntseg2=3
- ntseg1=26
- nbrp=0

so 

- tq = 1
- ntseq1+ntseg2+3=32
- product = 32*1=32

**my original sample**
- ntseg2=0
- ntseg1=1
- nbrp=7

so 
- tq = 8
- ntseq1+ntseg2+3=4
- product =4*8=32

But first sample - produces a lot of receive errors (counter is not zero) and my second samples works very well.
Did you have same problems in the past? How can I handle this problem?

alexanderwachter commented:
Well, you can set the sample point a bit earlier to meet your old values. However, the sample points used by Zephyr are recommended by the CiA

benys commented:
@alexanderwachter thank you.
I used to validate this harware: https://ucandevices.github.io/uccb.html so I understand that it propably set different params a litte not compatible".

So I push my changes related with only 1 interrupt line (instead of two).

benys commented:
I added changes related with bindings/boards to this PR.
For example in my board "counter" sample works.

henrikbrixandersen commented:
> I added changes related with bindings/boards to this PR.
> For example in my board "counter" sample works.

Your board needs to pass all the tests/samples tagged with `can`. Please run `twister --device-testing -t can ...` against your board and confirm that all tests are passing (you may or may not run into trouble with the ISO-TP tests, please report back if you do).

Also, please split these changes into multiple, logically separate commits. One for adding the compatible, one for adding the driver, one for adding SoC support, board support etc.

benys committed:
sam: can: CAN driver for SAM0 socs

Driver was based on can_sam. SAMC21 has only 1 interrupt for one
can "output", so can interrupt has to executes two lines of
interrupts.
CAN is configured to use OSC48M clock via GLCK7. GLCK7 is set
by divider configured from dts.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

benys committed:
sam: atsamc21: enable CAN driver for SAMC21

Enable CAN driver sam0 in SAMC21 socs. CAN module exists only in
C21 socs.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

benys committed:
sam: atsamc21: enable CAN module in c21 board

Configure pinmux pins at samc21 xpro board. Boards has
hard-configured pins for CAN0 and CAN1 module.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

benys commented:
@henrikbrixandersen I seperated commits in logically:
* one for driver and required changed in mcan
* one for soc support
* one for board configuration

Yes the twister passes, I forget mention it :-)

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,181 @@
+	/* Enable the GCLK */
+	GCLK->PCHCTRL[cfg->gclk_core_id].reg = GCLK_PCHCTRL_GEN_GCLK0 |
+					       GCLK_PCHCTRL_CHEN;
+	/* Enable SERCOM clock in MCLK */
```
SERCOM?

henrikbrixandersen reviewed (dismissed):
Looks good! One small observation, not a blocker.

str4t0m commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,181 @@
+	const struct can_mcan_config *mcan_cfg = dev->config;
+	const struct can_sam0_config *sam_cfg = mcan_cfg->custom;
+
+	*rate = SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / (sam_cfg->divider);
```
The divider seems to be only used for rate calculation, configuration of the divider is missing.

str4t0m commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,181 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <soc.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
```
Could you please also add `#include <zephyr/irq.h>` as you are using `IRQ_CONNECT`

str4t0m reviewed (changes requested):
A few comments.

henrikbrixandersen commented:
@str4t0m Please revisit.

str4t0m commented:
There are quite a few minor issues, so it would be great if @benys could address them and rebase on the latest main.

benys commented:
I will rebase ASAP :) 
In very near feature I have to come back to this PR:) beocuse I need it.

stephanosio commented:
> I will rebase ASAP :)
> In very near feature I have to come back to this PR:) beocuse I need it.

@benys FYI, the feature merge freeze for v3.3.0 is later today (27th January), so you will need to do it very quick if you want this feature in the upcoming release.

benys commented:
> > I will rebase ASAP :)
> > In very near feature I have to come back to this PR:) beocuse I need it.
> 
> @benys FYI, the feature merge freeze for v3.3.0 is later today (27th January), so you will need to do it very quick if you want this feature in the upcoming release.

I will try!

benys commented:
@stephanosio @str4t0m I fixed all issues - I hope :)

str4t0m commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+	const struct can_mcan_config *mcan_cfg = dev->config;
+	const struct can_sam0_config *sam_cfg = mcan_cfg->custom;
+
+	*rate = SOC_ATMEL_SAM0_GCLK0_FREQ_HZ / (sam_cfg->divider);
```
Thanks!
I am not really familiar with Atmel SoCs, but in `can_sam0_clock_enable()` you are setting `GCLK_GENCTRL_SRC_OSC48M` as clock source for the _clock generator 1_, which I assume is a fixed 48MHz oscillator. But here you are using `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` which depends on the selected sysclock frequency and might be changed. Shouldn't you set it to 48MHz or a macro representing the 48MHz here?

Also, the clock generator 1 is part of the Generic Clock Controller peripheral, so having it's divider as a property of the CAN is not ideal. Maybe @nandojve can weight in? Can we introduce that as a temporary solution until a generic clock controller driver is available for the series?

nandojve replied:
I'm not expert on CAN but as far I know clock should be very accurate. I think there is no problem to use `OSC48M` as base clock with proper factory (MCHP) calibration. What could be an issue is that DPLL is not enabled at this moment and that, on my view, should be the GCLKGEN reference for CAN. Usually a `XOSC[32K]` are a better reference for DPLL but you can use `OSC48M` -> `GLKC` -> `DPLL input ref` that will be far better than only `GLKC` (in theory).

Usually without a good clock base CAN may suffer when temperature changes, for instance.

henrikbrixandersen replied:
What is recommended clock source/frequency for the MCAN block according to the datasheet?

benys replied:
@henrikbrixandersen documentation doesn't specify recomended source of clock.

> Clocks
> An AHB clock (CLK_CAN_AHB) is required to clock the CAN. This clock can be configured in the Main Clock
peripheral (MCLK) before using the CAN, and the default state of CLK_CAN_AHB can be found in the
MCLK.AHBMASK register.
> A generic clock (GCLK_CAN) is required to clock the CAN. This clock must be configured and enabled in the generic
clock controller before using the CAN.
> This generic clock is asynchronous to the bus clock (CLK_CAN_AHB). Due to this asynchronicity, writes to certain

Datasheet only specify that MCAN has to be connected to GLCK.

str4t0m replied:
I my first comment I wanted to highlight that:
- You are using GLCK1 as clock source for the CAN peripheral.
- For rate calculation in line 46 you use GLCK0 frequency divided by the GLCK1 divider
- But GLCK1 is feed by OSC48M directly and not in series after GCLK0.
- This will work now as the system clock(GLCK0) is also fed by OSC48M and not divided. But as soon as that changes, chances are high that the CAN driver is overlooked and rate calculation is wrong.
- But if Atmel maintainers are fine with it I won't block.

henrikbrixandersen replied:
@nandojve Just to make sure - are you okay with the above limitations?

attie-argentum replied:
I'm also not convinced that `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` is appropriate here...

benys replied:
I changed reference from `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` to `SOC_ATMEL_SAM0_OSC48M_FREQ_HZ`
I hope that it is ok for you :)

henrikbrixandersen commented on boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:
```diff
@@ -151,3 +152,29 @@
+	pinctrl-0 = <&can0_default>;
+	pinctrl-names = "default";
+
+	bus-speed = <1000000>;
```
```suggestion
	bus-speed = <125000>;
```

benys replied:
I don't understand :( i changed it as i receive review from other person.

Have i to change it?

str4t0m replied:
Yes please. The only data phase(CAN-FD) should have the higher bit rate of 1MHz.

henrikbrixandersen commented on boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:
```diff
@@ -151,3 +152,29 @@
+	pinctrl-0 = <&can1_default>;
+	pinctrl-names = "default";
+
+	bus-speed = <1000000>;
```
```suggestion
	bus-speed = <125000>;
```

henrikbrixandersen commented on dts/bindings/can/atmel,sam0-can.yaml:
```diff
@@ -0,0 +1,29 @@
+
+compatible: "atmel,sam0-can"
+
+include:
+  - name: can-fd-controller.yaml
+  - name: pinctrl-device.yaml
```
Aren't you missing an include on `bosch,m_can-base.yaml`?

benys replied:
In my opinioin - no - because `bosch,m_can-base.yaml` is defined on controller level, not in individual module.
The same way is defined for sam-can.

str4t0m reviewed (commented):
Thanks for the update.
Most issues are resolved, but I am not sure about configuring a GCLK channel divider via a property of the CAN node.

benys commented:
> Thanks for the update.
> Most issues are resolved, but I am not sure about configuring a GCLK channel divider via a property of the CAN node.

I had the same issue. I checked how it was implemented at armel SAM - it was the same pattern.

For now c21 doesn't use GLCK other then 0.

nandojve commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+static void can_sam0_clock_enable(const struct can_sam0_config *cfg)
+{
+	/* Enable the GLCK1 with DIV*/
+	GCLK->GENCTRL[1].reg = GCLK_GENCTRL_SRC(GCLK_GENCTRL_SRC_OSC48M)
+				| GCLK_GENCTRL_DIV(cfg->divider)
+				| GCLK_GENCTRL_GENEN;
```
```diff
+	GCLK->GENCTRL[1].reg = GCLK_GENCTRL_SRC(GCLK_GENCTRL_SRC_OSC48M)
+			     | GCLK_GENCTRL_DIV(cfg->divider)
+			     | GCLK_GENCTRL_GENEN;
```

benys replied:
I am sorry that forget this tyle. I fixed :)

nandojve commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+				| GCLK_GENCTRL_GENEN;
+
+	/* Route channel */
+	GCLK->PCHCTRL[cfg->gclk_core_id].reg = GCLK_PCHCTRL_GEN_GCLK1 |
+					       GCLK_PCHCTRL_CHEN;
```
```diff
+	GCLK->PCHCTRL[cfg->gclk_core_id].reg = GCLK_PCHCTRL_GEN_GCLK1
+					     | GCLK_PCHCTRL_CHEN;
```

nandojve commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+		.phase_seg1 = 0x01,
+		.phase_seg2 = 0x01,
+		.prescaler = 0x01
+		},
+	.timing_max = {
```
Keep style...fix all entries

```diff
-		},
-	.timing_max = {
+	},
+	.timing_max = {
```

nandojve commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+	} while (false)
+
+#define CAN_SAM0_IRQ_CFG_FUNCTION(inst)                                                         \
+static void config_can_##inst##_irq(void)                                                      \
+{                                                                                              \
+	SAM0_CAN_IRQ_CONNECT(inst, 0);                                            \
+}
+
+#define CAN_SAM0_CFG_INST(inst)						\
+	static const struct can_sam0_config can_sam0_cfg_##inst = {		\
+		.mclk = (volatile uint32_t *)MCLK_MASK_DT_INT_REG_ADDR(inst),	\
+		.mclk_mask = BIT(DT_INST_CLOCKS_CELL_BY_NAME(inst, mclk, bit)),	\
+		.gclk_core_id = DT_INST_CLOCKS_CELL_BY_NAME(inst, gclk, periph_ch),\
+		.divider = DT_INST_PROP(inst, divider),			\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),		\
+		.config_irq = config_can_##inst##_irq,		\
+	};						\
+									\
+	static const struct can_mcan_config can_mcan_cfg_##inst =	\
+		CAN_MCAN_DT_CONFIG_INST_GET(inst, &can_sam0_cfg_##inst);
+
+#define CAN_SAM0_DATA_INST(inst)						\
+	static struct can_sam0_data can_sam0_data_##inst;			\
+									\
+	static struct can_mcan_data can_mcan_data_##inst =		\
+		CAN_MCAN_DATA_INITIALIZER(&can_sam0_data_##inst.msg_ram, \
+					  &can_sam0_data_##inst);	\
+
+#define CAN_SAM0_DEVICE_INST(inst)					\
+	DEVICE_DT_INST_DEFINE(inst, &can_sam0_init, NULL,		\
+			      &can_mcan_data_##inst,			\
+			      &can_mcan_cfg_##inst,			\
+			      POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,	\
+			      &can_sam0_driver_api);
+
+#define CAN_SAM0_INST(inst)                                                     \
+	PINCTRL_DT_INST_DEFINE(inst);                                          \
+	CAN_SAM0_IRQ_CFG_FUNCTION(inst)                                         \
+	CAN_SAM0_CFG_INST(inst)                                                 \
+	CAN_SAM0_DATA_INST(inst)                                                \
```
Keep `\` vertically align.
Use only TABs for this purpose
Loop longest line and add 1 TAB+\
Then repeat for the others.
Zephyr is 8 space TAB like Linux.

```diff
#define SAM0_CAN_IRQ_CONNECT(inst, m)							\
...
+		.gclk_core_id = DT_INST_CLOCKS_CELL_BY_NAME(inst, gclk, periph_ch),	\
		.divider = DT_INST_PROP(inst, divider),					\
		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),				\
		.config_irq = config_can_##inst##_irq,					\
	};										\
...
#define CAN_SAM0_DEVICE_INST(inst)							\
	DEVICE_DT_INST_DEFINE(inst, &can_sam0_init, NULL,				\
			      &can_mcan_data_##inst,					\
			      &can_mcan_cfg_##inst,					\
			      POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,			\
			      &can_sam0_driver_api);
...
#define CAN_SAM0_INST(inst)								\
	PINCTRL_DT_INST_DEFINE(inst);							\
	CAN_SAM0_IRQ_CFG_FUNCTION(inst)							\
	CAN_SAM0_CFG_INST(inst)								\
	CAN_SAM0_DATA_INST(inst)							\
```

nandojve reviewed (changes requested):
Overall, LGTM. I only checked a few styles that are used for these series.

I let clock decision for CAN folks. I assume that HW works and should be enough to introduce the feature. Maybe a better commit message that advice that clock should be improved to DPLL may help users.

nandojve commented:
I took liberty to add removed persons from review list because I think it was by mistake.

benys commented:
> I took liberty to add removed persons from review list because I think it was by mistake.

Yes it was by mistake. I didn't know how revert my mistake.

benys commented:
@henrikbrixandersen I would like to back to this this PR :)
Should I rebase to newest one? Need I do more?

str4t0m reviewed (changes requested):
Yes please rebase on the latest main.

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+
+	ret = pinctrl_apply_state(sam_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
```
Please either add logging of failures or remove the log declaration. I would prefer the former.

henrikbrixandersen reviewed (commented):
LGTM, pending response from @nandojve.

nandojve reviewed (approved):
LGTM!

attie-argentum commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,188 @@
+static void can_sam0_clock_enable(const struct can_sam0_config *cfg)
+{
+	/* Enable the GLCK1 with DIV*/
+	GCLK->GENCTRL[1].reg = GCLK_GENCTRL_SRC(GCLK_GENCTRL_SRC_OSC48M)
+			     | GCLK_GENCTRL_DIV(cfg->divider)
+			     | GCLK_GENCTRL_GENEN;
```
All other SAM parts put this level of clock configuration in the SOC's code, e.g: [`.../soc_samc2x.c`](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samc2x.c)

Given that `cfg->divider` comes from each channel in the device tree, I'm also concerned that in applications that use both channels, there could be a silent incompatibility between the two... the device tree currently suggests that these dividers are independent, when they aren't!

I note that the f<sub>GCLK_CANn</sub> is listed as "_48 MHz max_", and the prescaler in the CAN peripheral is fairly limited, so is probably required...

I'm concerned that placing the GCLK1 configuration in the driver will make it less visible and potentially introduce issues in the future...
- At minimum, please add a comment to `soc_samc2x.c` to indicate that GCLK1 is reserved for CAN
- My preference would be to move configuration of GCLK1 into `soc_samc2x.c`
- Consider using two GCLK channels (one for each CAN interface), or perhaps move the `divider` value up into the `can` / `bosch,m_can-base` node

e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/ed2c25effccc55e459e2b96f389d9b1a177daf7d/soc/arm/atmel_sam0/common/soc_saml2x.c#L26-L29

attie-argentum replied:
Ah, well found... Using `GCLK0` @ 48 MHz sounds sensible to me...

benys replied:
@attie-argentum  @nandojve 
About clocks I inspired with can_sam and can_mcan.
In my project I need divider to set valid ntseg (i had to set specific ntseg1/ntseg2 values for specific clock speed).

Can I change can0 to config GLCK7 instead of GLCK1? It will be ok for you all?

nandojve replied:
You mean reserve the GCLK[7] for CAN? I don't see problems with that but you need make sure all SAM0 series enable GCLK[7] at soc. Is that what you want to do, right?

benys replied:
@nandojve as you can see I "dynamicly" enable at cansam0 driver. So that it isn't need to to enable at soc.
BTW In my knowlage CAN module for sam0 has only samC21 series.

nandojve replied:
Ah ok! But as a remind that ATSAME5x have CAN-FD too.
Fine for me if you want move from GCLK1 to GCLK7.

benys replied:
I believe that "conversation" can be resolved :-)
Thanks for info about E5x, I didn't know.

attie-argentum reviewed (dismissed):
Thanks @benys ... aside from the small stylistic one, I have concerns about the clock configuration. Please see my comments.

benys commented:
I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have overlay to configure test.
Should I here add overlay file?

nandojve commented:
> I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have exluding for this board...
>
> What next :-)?

I think there is no difference on drivers so maybe you can just exclude it. There is no point to build the exactly same driver if there is no changes between device series that it is important catch. You can achieve that creating another commit and put it as the first on this series. So, you fix the build problem and introduce new driver.

benys commented:
> > I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have exluding for this board...
> > What next :-)?
> 
> I think there is no difference on drivers so maybe you can just exclude it. There is no point to build the exactly same driver if there is no changes between device series that it is important catch. You can achieve that creating another commit and put it as the first on this series. So, you fix the build problem and introduce new driver.

I created new PR - https://github.com/zephyrproject-rtos/zephyr/pull/56487 but on my computer builds normally. In twister I see that it try to use overlay not from board but only for seggerrtt

benys commented:
@attie-argentum & @str4t0m can you checkPR :-)?
I hope that is OK for you!

str4t0m reviewed (dismissed):
Thanks for addressing my concerns.
One note but not blocking.

henrikbrixandersen commented:
@attie-argentum Please revisit.

silverv commented:
Is it possible to use CAN1 and were there tests for it?

attie-argentum commented:
> @attie-argentum Please revisit.

@henrikbrixandersen - apologies for my delay. I've dismissed my review - it looks like my comments have been resolved, and I don't have the time to formally re-review right now. Others have approved, so please continue!

fabiobaltieri commented:
@benys can you rebase please?

benys commented:
> @benys can you rebase please?

I rabased, but I will it test +- tommorow at the office :) I will inform you.

silverv commented:
My experience so far is that with commit 81afde13448808f71b12c4bacc18d9ef4f6649ec, both CAN0 and CAN1 initialized but CAN1 gets stuck at send on a atsamc21g18a. As of [782c99f](https://github.com/zephyrproject-rtos/zephyr/pull/50408/commits/782c99fbdea65453b8173d9db1b787891b06fbba), atsamc21n18a, neither CAN peripheral would get past mcan initialization, I will double check everything now.

henrikbrixandersen commented on drivers/can/can_sam0.c:
```diff
@@ -0,0 +1,225 @@
+
+#define CAN_SAM0_CFG_INST(inst)								\
+	static const struct can_sam0_config can_sam0_cfg_##inst = {			\
+		.mclk = (volatile uint32_t *)MCLK_MASK_DT_INT_REG_ADDR(inst),		\
```
```suggestion
                .base = (mm_reg_t)DT_INST_REG_ADDR(inst),               \
		.mclk = (volatile uint32_t *)MCLK_MASK_DT_INT_REG_ADDR(inst),		\
```

henrikbrixandersen replied:
Where are you assigning `config->base`?

benys commented:
@silverv @henrikbrixandersen  i little fixed my PR, but after rebase still not works :-(
I am still looking reason...

benys committed:
can: mcan: fix setup configuration ram

SAM0 required can module was in init state before configure
pointers to ram which handle can frames.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

henrikbrixandersen reviewed (approved):
Thanks!
